scheduler/src/cook/mesos/optimizer.clj

```diff
@@ -19,122 +19,251 @@
+     A schedule: a map where the keys are milliseconds in the future
+     and  the values are recommendations to take at that point in the future"))
 
 (defn create-dummy-optimizer
   "Returns an instance of Optimizer which returns an empty schedule"
   [_]
   (log/info "Creating dummy optimizer")
   (reify Optimizer
-    (produce-schedule [this queue running available host-infos]
-      {0 {:suggested-matches {}}})))
+    (produce-schedule [_ _ _] {0 {:suggested-matches []}})))
 
-(defn optimizer-cycle!
+(defn instance-runtime
+  "Returns the instance runtime."
+  [instance now]
+  (let [start (:instance/start-time instance)
+        end (:instance/end-time instance now)]
+    (- (.getTime end)
+       (.getTime start))))
+
+(defn calculate-expected-complete-time
+  "Returns the expected time remaining to job completion.
+   Simple model for if run time exceeds expected runtime: the more it exceeds expected, the longer we expect it to take.
+   TODO: get better model!"
+  [instance expected-duration now]
+  (->> (instance-runtime instance now)
+       (- expected-duration)
+       Math/abs))
+
+(defn job->batch-uuid
+  "Retrieves a batch UUID for the provided job."
+  [job]
+  (or (-> job :group/_job :group/uuid)
+      (:job/uuid job)))
+
+(defn waiting-job->optimizer-job
+  "Converts a waiting job to an optimizer job representation."
+  [job default-runtime]
+  (let [expected-duration (get job :job/expected-runtime default-runtime)
+        resources (util/job-ent->resources job)
+        batch-uuid (job->batch-uuid job)]
+    (merge {:batch batch-uuid
+            :count 1
+            :dur expected-duration
+            :expected_complete_time -1
+            :running_host_group -1 ; TODO write function to determine this
+            :state "waiting"
+            :submit_time (-> job :job/submit-time .getTime)
+            :user (:job/user job)
+            :uuid (:job/uuid job)}
+           resources)))
+
+(defn running-instance->optimizer-job
+  "Converts a running instance to an optimizer job representation."
+  [instance default-runtime now host-name->host-group]
+  (let [job (:job/_instance instance)
+        expected-duration (get job :job/expected-runtime default-runtime)
+        expected-complete-time (calculate-expected-complete-time instance expected-duration now)
+        resources (util/job-ent->resources job)
+        batch-uuid (job->batch-uuid job)]
+    (merge {:batch batch-uuid
+            :count 1
+            :dur expected-duration
+            :expected_complete_time expected-complete-time
+            :running_host_group (-> instance :instance/hostname host-name->host-group)
+            :state "running"
+            :submit_time (-> job :job/submit-time .getTime)
+            :user (:job/user job)
+            :uuid (:job/uuid job)}
+           resources)))
+
+(defn create-optimizer-server
+  [{:keys [cpu-share default-runtime host-info max-waiting mem-share opt-params opt-server step-list]
+    :or {opt-params {}}
```

Comment:
Do we need a docstring here? E.g., documentation on the `opt-params` might be useful.